ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/NarayanaTransactionManager.java

```diff
@@ -319,6 +333,10 @@ public void begin() throws NotSupportedException, SystemException {
    */
   @Override
   public void commit() throws HeuristicMixedException, HeuristicRollbackException, RollbackException, SystemException {
+    if (this.transactionScopeBeforeDestroyedBroadcaster != null) {
+      this.transactionScopeBeforeDestroyedBroadcaster.fire(this.getTransaction());
```

Comment:
This is only to fire the scope event. The code may observe that event (see the test `@Observes @BeforeDestroyed`). This code does not run the destruction, it just informs that the scope will be destroyed.
The spec is here
https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html#builtin_contexts
I wrote a bit about it at the blog http://jbossts.blogspot.com/2019/04/jta-and-cdi-integration.html
you can read the discussion in the forum from the issue https://issues.redhat.com/browse/JBTM-3106

Reply:
It's a bit similar to beforeCompletion but rather from the user perspective. This is about transaction scoped of the CDI spec. The `beforeCompletion` is about the synchronization from JTA spec.

Reply:
Ah yes, I do recall that blog post now, it was very informative.

The [spec says that the transaction scope is active](https://github.com/eclipse-ee4j/jta-api/blob/initial-contribution/src/main/java/javax/transaction/TransactionScoped.java#L31) when the return from a call to `UserTransaction.getStatus` is one of the following states:
```
 * <li>Status.STATUS_ACTIVE
 * <li>Status.STATUS_MARKED_ROLLBACK
 * <li>Status.STATUS_PREPARED
 * <li>Status.STATUS_UNKNOWN
 * <li>Status.STATUS_PREPARING
 * <li>Status.STATUS_COMMITTING
 * <li>Status.STATUS_ROLLING_BACK
```
If you broadcast the destroy scope event before the commit call then you will miss the STATUS_PREPARED, STATUS_PREPARING, STATUS_COMMITTING and STATUS_ROLLING_BACK states.

Reply:
Sure, but we are not about to fire an event on changing transaction state. The event is fired once before the scope is destroyed (i.e. `an event with qualifier @BeforeDestroyed(X.class) when a custom context is about to be destroyed`).
The active state is reflected in method checking the state https://github.com/jbosstm/narayana/blob/b4d4b04cad2e5ca376af0f753bc2c3cd6b097a0b/ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/TransactionContext.java#L143

Reply:
The spec says that the scope remains active until the transaction is about to transition into one of the end states.
When you broadcast the pre destroy event before the commit call the transaction will transition through multiple states before the transaction is regarded as inactive.
I'd like to see the rationale for broadcasting here as opposed to all of the other states where you could have broadcasted it (for example, what would stop you from generating the event immediately after the transaction is created, or after participants have prepared, or ... etc)?

Reply:
The rationale to broadcast event from here are two. This is the CDI event which belongs to CDI code. This is the place where the emitting such event fits the best. Second, this CDI event when applied to JTA, from my CDI and JTA specs interpretation, is what the `Synchronization.beforeCompletion` is processed. The `beforeCompletion` is handled in the same/similar time and thus the code is placed here.

Reply:
Broadcasting the event when there are still other states to transition through seems like an arbitrary choice.
Can you point me to where the spec says it is valid to generate it when the transaction state is `Status.ACTIVE`.

Reply:
It may be the case that whoever did the spec did not think this through completely. The before destroy bean event does not seem to add any value and is potentially quite difficult to implement. Paul may have some input.

Reply:
If it doesn't add value and you find that it is under specified then one could then argue that it doesn't really matter when the event is generated since nobody will be relying on it. Is there a CDI TCK we could check (if so it might tell us the expectations).

Reply:
@mmusgrov I discussed this with the CDI/Weld team (namely with @manovotn) and he agreed that it would be better fitting to emit the event just before the context is destroyed. I'm now trying to find if there is a way to place it more properly. I will start with Quarkus code (that is a bit different to what we have in Narayana but it's pretty similar) - https://github.com/quarkusio/quarkus/issues/17658. If the processing is agreed I will move the concept to the Narayana code. If there is a real troubles with adjusting the current behaviour I will report back to you.

`It may be the case that whoever did the spec did not think this through completely.` Agree and Matej agreed as well. The custom context event observers are pretty corner cases. The CDI spec considers mostly the core CDI contexts.

Reply:
> It may be the case that whoever did the spec did not think this through completely. The before destroy scope event does not seem to add any value and is potentially quite difficult to implement.

Indeed this is very rarely used which is even more true for custom contexts (any non-builtin). However, it should not be difficult to implement. Those are simple CDI events fired around context activation/deactivation. By that you should understand (from CDI point of view) thresholds when user requesting a bean with certain scope gets the bean versus gets `ContextNotActiveException`.

Reply:
@mmusgrov after quite an extensive investigation of the topic. In summary the current PR is pretty good to what we need to accomplish with these events.

Implementation details:
The transaction scope is bound to the existence of the transaction. When it's created - after begin - then the scope needs to be active. Then the before destroyed event is really being fired at time when transaction is active (as that's the time the txn context is still available) but when it's coming time to be destroyed. The JTA API does not provide any other way than `beforeCompletion` for such hook. It's exactly the time before the `commit/rollback` is processed. I cannot find another good place where to put the event to be fired.
I experimente with some options around context management (https://github.com/ochaloup/narayana/compare/master...JBTM-3416-before-destroy-event-new) but I ended with the same result. On top of that the `TransactionScopeCleanup#beforeCompletion` hook cannot be used as the context bean is created only when some `@ScopedBean` exitst. But the events need to be fired for any transaction - i.e. for empty context as well. Then the current location of the event management is the proper place.

Can you, please, re-review the  PR and consider if this is fine to be done this way or propose a concrete idea how to hook a CDI eventing in a different way? If there is some ambiguity I can try to explain or even better we can organize a call where we can discuss all the details.

Reply:
> I cannot find another good place where to put the event to be fired.

This statement implies that because you cannot implement it correctly you are choosing a semi random point in the JTA transaction lifecycle to emit the before CDI scope destroyed event (and to be honest I would rather we just not support the event if we are unable to implement it correctly). @manovotn suggested that "it should not be difficult to implement", Matej are you able to provide us with any help on this problem?

Reply:
> @manovotn suggested that "it should not be difficult to implement"

I merely meant to implement this CDI behavior from just about anywhere you need in your code is simple (e.g. you can fire the event even via bean manager).

But me and @ochaloup had a call yesterday where we went over this problem. Ondra explained how your context works and that it is lazily initialized. However, these events should be fired even for empty context; they are an indication of context state/readiness, not its contents. With that in mind and given the lazy context init you are using, the current solution seems to be the best you can achieve.

>  (and to be honest I would rather we just not support the event if we are unable to implement it correctly)

These events are a connecting membrane between CDI itself code logic of your context which makes the "correct way" different for every context. I think the current solution is actually good. CDI requirements on these events basically boil down to:
* Firing `@Initialized` when the context is activated (`Context.isActive()` returns true) and within this event user can resolve transactional scoped beans
* Firing `@BeforeDestroyed` some time before shutting it down, the context still needs to be active and user has to be able to resolve transactional scoped beans - think of this as the last chance for user to reliably operate within active transactional context
* Firing `@Destroyed` once you torn down whole context, bean with transactional scope can no longer be resolved (user would get `ContextNotActiveException`)

Reply:
Thanks @manovotn .

@mmusgrov I'm sorry but really I need to object to the statement `you are choosing a semi random point`. To my best knowledge this is the best and correct place to emit the event.

Reply:
I already indicated that generating the event misses multiple states after transaction leaves the state `Status.ACTIVE` when the context should still be active. Ideally the destroy event needs to be fired when the TM has received all the responses back after the prepare phase (but even that misses STATUS_COMMITTING, in which case we'd need to wait for all the commit responses to arrive).

To be honest, the spec has not addressed the needs of JTA correctly and this behaviour is probably unimplementable.

@manovotn the spec says that the context remains active whenever a call to UserTransaction.getStatus returns any of

```java
 * <li>Status.STATUS_ACTIVE
 * <li>Status.STATUS_MARKED_ROLLBACK
 * <li>Status.STATUS_PREPARED
 * <li>Status.STATUS_UNKNOWN
 * <li>Status.STATUS_PREPARING
 * <li>Status.STATUS_COMMITTING
 * <li>Status.STATUS_ROLLING_BACK
 ```
 
The PR fires the event when the status is STATUS_ACTIVE. But the context will/should remain active for all of the subsequent states. Firing it when the status is ACTIVE doesn't add much, you could even fire the event immediately after the transaction is created which would not be of use to anyone.

I can withdraw my review and let someone else do it, but the implementation in this PR is just choosing an arbitrary point from which to generate the event.

Reply:
> To be honest, the spec has not addressed the needs of JTA correctly

Well, it is not tailored to JTA. It is a generic statement for all custom contexts which can be provided by other specs or even by users.

> The PR fires the event when the status is STATUS_ACTIVE. But the context will/should remain active for all of the subsequent states. Firing it when the status is ACTIVE doesn't add much, you could even fire the event immediately after the transaction is created which would not be of use to anyone.

Talking about `@Initialized` and `@BeforeDestroyed`, those are still useful in just the same way. I can see the dispute between what JTA requires and how you want it to work with CDI but I think you might be a bit too punctual. For a CDI user, current behavior is still handy - it notifies you at a time where you know you can rely on the scope being active (Initialized) and gives you a chance to perform cleanup on your beans before JTA internal logic takes over (BeforeDestroyed) and does its job in finishing the transaction and destroying the context. It doesn't necessarily need to be the exact very first/last moment of tx at which you fire those events for them to have value to users.

Reply:
So @manovotn are you saying that if I dismiss my review you will commit to reviewing it?

Reply:
Actually I am not marked as a reviewer. Therefore since I don't think this requirement is implementable I won't review it. @ochaloup will you assign @manovotn as reviewer. @mayankkunwar I think you are also marked as a reviewer so will you either provide a review or decline the review request, thanks.

Reply:
Sure, I can do that. Just bear in mind that my expertise lies mainly in CDI area, not so much in JTA :-)

Reply:
@ochaloup the PR changes are good according to the current CDI implementation(which is generic and not according to JTA specs) , but I think @mmusgrov has a valid point. This would not be the right place for firing the @BeforeDestroy event, maybe somehow if we could manage to place it just before transaction cleanup?

Reply:
Yes the spec has left some scope for different interpretations of when to fire the event. For example one could generate the before destroy context event as soon as the transaction begin is called and still be within the letter of the spec.

Reply:
Maybe we could keep an eye on this one, if we could raise a request to CDI team to handle this scenario in upcoming versions, if that is possible?